sympy/core/function.py

```diff
@@ -871,7 +884,7 @@ def __new__(mcl, name, bases=(AppliedUndef,), __dict__=None, **kwargs):
         # Save these for __eq__
         __dict__.update({'_extra_kwargs': kwargs})
         __dict__['__module__'] = None # For pickling
-        ret = super(UndefinedFunction, mcl).__new__(mcl, name, bases, __dict__)
+        ret = super(UndefinedFunction, mcl).__new__(mcl, str(name), bases, __dict__)
```

Comment:
Isn't the case that `name` is always a string? As far as I know they are used for printing, right? Is this required for your changes?

Reply:
It will not be a string when called with `Function(Symbol('f'))`, which is partly what this PR aims to allow. Supplying `Symbol` in place of a string in `Function` constructor (the desired functionality) will not work without some workaround, and this was the simplest way I could come up with.

I guess an argument can be made that because `Function(Symbol('f'))` was not possible before, then this new functionallity should only be applied to `IndexedBase`, leaving `Function` be as it is. Thoughts?

Reply:
I will read the thread here and the related PRs, will try out some experiments.  I will present my views after that. Thanks for the explanation. The work seems interesting.

Reply:
The story is that `Symbol('f')(x)` formerly automagically created `Function('f')(x)` but this was frowned upon. The current PR would allow `Function(Symbol(f))` which is about as painless a way to go from Symbols to Functions while retaining the same assumptions.

So what is the use case for this? One would almost always include and argument for the function, e.g. `x -> x(t) ` so to make that change would be either `x.subs(x, Function(x.name, x.assumptions0)(t))` or `x.subs(x, Function(x)(t))`. Are there other use cases? Is there a compelling reason to not allow creation of a a Function from an atom (Symbol) with assumptions?

Reply:
Thanks for providing the summary. I think then this change is fine. I am supportive of this PR. For a simpler way than this I will think and if needed change can be made in future sometime.